indico/web/client/js/custom_elements/ind_date_picker.js

```diff
@@ -31,6 +31,11 @@ customElements.define(
       return this.getAttribute('end');
     }
 
+    get invalid() {
```

Comment:
Not sure if there's a better way to indicate that the field is invalid, since the `value` is only set in case the field has a valid value.

Reply:
Will take a look tomorrow. This is related to the comment in the PR, yes?

Reply:
Nope, unrelated. This PR is just to have a generic way of using the DatePicker in any final-form-based form (outside the registration form).

Reply:
The custom element gives you two values to play with. There's the raw value `IndDatePicker.value` and a parsed value `IndDatePicker.date`. If I'm understanding what you are trying to do with `invalid`, you should be able to do something like this:

```javascript
  function handleDateChange(ev) {
    const {date, value} = ev.target.closest('ind-date-picker');
    const invalid = !!value && !date;
    // en-CA uses the ISO format date (YYYY-MM-DD) but gives it in local time zone.
    // Do not use toISOString() for this because it may result in incorrect date due
    // to time zone differences.
    onChange(invalid ? INVALID : date?.toLocaleDateString('en-CA'));
  }
```

Reply:
The parser always returns `undefined` if it cannot reasonably parse a date string with the given format string.

Reply:
Ah indeed, `this.date` also exists, I didn't notice it since it's being defined during initialization and not on the class scope. So my `get invalid` method can be simplifed:

```
    get invalid() {
      return !!this.value && !this.date;
    }
```

(I prefer having the property in there, since that way it's all in one place)

Reply:
To me `this.value` and `this.date` do not belong in the same bucket as `invalid` because the former two represent the value of the input, while the latter represents the interpretation of the value from the business standpoint. In other words, `invalid` is a higher-level concept. 

Now, it's true that validation (which is what this basically is) can be built into  an input in the DOM API, but if we're gonna do that, I'd prefer to do it in a standard way (Constraints validation API), and then use that. As that opens doors like fully embracing the built-in validation facilities and even potentially getting rid of FinalForm.

Reply:
Sounds reasonable, I moved the `invalid` logic into the React component.

I guess w/ the constraints validation API you mean `setCustomValidity` on the DOM level? Anyway, I don't see final-form going away since it's very useful for react-based forms. ;)

Reply:
> Anyway, I don't see final-form going away since it's very useful for react-based forms. ;)

To me, there's no difference between vanilla forms and React forms. React simply renders the forms. It doesn't *have* to manage the form state (already managed by the browser), nor the states of individual controls (React is actually horrible for that). The only things you actually need to handle on the React side is:

- specify what controls get rendered
- transform the data for submission

Most of the form complications in React come from trying to make React do things it doesn't have to do. Final forms is just a patch for such issues. Or, as I like to say, solution to a problem that shouldn't have been a problem to begin with.